jena-fuseki2/jena-fuseki-mod-geosparql/src/test/java/org/apache/jena/fuseki/mod/geosparql/TestFMod_SpatialIndexer.java

```diff
@@ -0,0 +1,147 @@
+
+public class TestFMod_SpatialIndexer {
+    private WebDriver driver;
+    private JavascriptExecutor js;
+
+    private static Envelope queryEnvelope = new Envelope(-180, 180, -90, 90);
+    private DatasetGraph dsg;
+    private SpatialIndex spatialIndex;
+    private Node graphName1 = NodeFactory.createURI("http://www.example.org/graph1");
+
+    @Before
+    public void setUp() throws IOException, SpatialIndexException {
+        dsg = DatasetGraphFactory.create();
+        setupTestData(dsg);
+
+        spatialIndex = SpatialIndexUtils.buildSpatialIndex(dsg);
+
+        String[] argv = new String[] { "--empty" };
+        FusekiServer server = FusekiMain.builder(argv)
+            .add("test", dsg)
+            .build();
+        server.start();
+        int port = server.getPort();
+        String serverURL = "http://localhost:" + port + "/";
+        String siteUrl = serverURL + "test/spatial";
+
+        ChromeOptions options = new ChromeOptions();
+        options.addArguments("--headless=new"); // use "new" headless mode (better support)
+        options.addArguments("--no-sandbox");
+        options.addArguments("--disable-dev-shm-usage");
+
+        WebDriverManager.chromedriver().setup(); // Automatically downloads and sets path
+        driver = new ChromeDriver(options);
+        driver.get(siteUrl);
+        js = (JavascriptExecutor) driver;
+    }
+
+    private void setupTestData(DatasetGraph dsg) {
+        // Fill the graph with a few geometries; spatial index construction will derive the SRS from them.
+        Envelope envelope = new Envelope(-175, 175, -85, 85);
+
+        Map<GeometryType, Number> conf = new HashMap<>();
+        conf.put(GeometryType.POINT, 1);
+
+        // Generate geometries into the default graph and a named graph
+        GeometryGenerator.generateGraph(dsg.getDefaultGraph(), envelope, conf);
+
+        conf.put(GeometryType.POLYGON, 1);
+        GeometryGenerator.generateGraph(dsg.getGraph(graphName1), envelope, conf);
+    }
+
+    @Test
+    @Ignore // Ignored for new because this test requires a local Chrome browser.
+    public void testIndexAndCleanButtons() {
+        // Index the test data (default graph and 1 named graph).
+        WebElement button = driver.findElement(By.id("apply-action"));
+        button.click();
+        awaitEvent();
+        Assert.assertEquals(1, spatialIndex.query(queryEnvelope, Quad.defaultGraphIRI).size());
+        Assert.assertEquals(2, spatialIndex.query(queryEnvelope, graphName1).size());
+        clearLastEvent();
+
+        // Remove the named graph and update the index.
+        dsg.removeGraph(graphName1);
+        WebElement cleanButton = driver.findElement(By.id("clean-action"));
+        cleanButton.click();
+        awaitEvent();
+        try (Stream<?> s = Iter.asStream(dsg.listGraphNodes())) {
+            long actualNamedGraphCount = s.count();
+            Assert.assertEquals(0, actualNamedGraphCount);
```

Comment:
Whoops, this should test whether the spatial index (not dsg) only has one graph.

Reply:
updated the test.